### ISSUE TITLE
We should set the PVC's volumeName, i think

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/kubespawner#274]
@hans-permana found the z2jh setting `singleuser.storage.dynamic.volumeNameTemplate`. We both thought that this could influence the naming of the dynamically created `PersistantVolume` objects. Our investigation led us to conclude that this is only the spawned pods volume reference under the pods `spec.volumes` that is named in this manner, while the actual `PersistantVolume` k8s object, which is to be provisioned by the cloud provider looking at the PVC object in turn created by KubeSpawner, is not getting a similar name. This is because the PVC kubespawner creates does not set the `spec.volumeName` field ([see reference documentation](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#persistentvolumeclaimspec-v1-core)).

I think we should allow optionally setting the `volumeName` field of the PVC's spec through the `make_pvc` function in kubespawner. The Z2JH project can then later decide to utilize this option or not.

## Suggested action - Add config of `volumeName` here
https://github.com/jupyterhub/kubespawner/blob/481ec0926960927300c7355c48bd03230ee857f0/kubespawner/objects.py#L393-L433

#### Z2JH configuration ([inspect in the source code](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/4075094ba2dbffa031a111e308134e33f4956a7b/jupyterhub/values.yaml#L200-L204)):

```yaml
singleuser:
  storage:
    dynamic:
      storageClass:
      pvcNameTemplate: claim-{username}{servername}
      volumeNameTemplate: volume-{username}{servername}
      storageAccessModes: [ReadWriteOnce]
```

#### Z2JH config of kubespawner based on Z2JH config:
```python
if storage_type == 'dynamic':
    pvc_name_template = get_config('singleuser.storage.dynamic.pvcNameTemplate')
    c.KubeSpawner.pvc_name_template = pvc_name_template
    volume_name_template = get_config('singleuser.storage.dynamic.volumeNameTemplate')
    c.KubeSpawner.storage_pvc_ensure = True
    set_config_if_not_none(c.KubeSpawner, 'storage_class', 'singleuser.storage.dynamic.storageClass')
    set_config_if_not_none(c.KubeSpawner, 'storage_access_modes', 'singleuser.storage.dynamic.storageAccessModes')
    set_config_if_not_none(c.KubeSpawner, 'storage_capacity', 'singleuser.storage.capacity')

    # Add volumes to singleuser pods
    c.KubeSpawner.volumes = [
        {
            'name': volume_name_template,
            'persistentVolumeClaim': {
                'claimName': pvc_name_template
            }
        }
    ]
    c.KubeSpawner.volume_mounts = [
        {
            'mountPath': get_config('singleuser.storage.homeMountPath'),
            'name': volume_name_template
        }
    ]
```

#### A spawned user pods specs
```yaml
spec:
  # ...
  volumes:
  - name: volume-erik-2esundell
    persistentVolumeClaim:
      claimName: claim-erik-2esundell
```

#### Docs about the `volumes` configurable on kubespawner

https://github.com/jupyterhub/kubespawner/blob/481ec0926960927300c7355c48bd03230ee857f0/kubespawner/spawner.py#L586-L611

#### Kubespawners actual usage of the volumes

https://github.com/jupyterhub/kubespawner/blob/481ec0926960927300c7355c48bd03230ee857f0/kubespawner/objects.py#L311-L312

**Comment by hans-permana** [comment]
This would also be useful to set the name of `hub-db-dir` pv. This is to prevent the creation of a new volume after each cluster creation.

**Comment by bitnik** [comment]
I just tried to create a PVC with `volumeName` set:
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: claim-test-volume-name
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
  volumeName: volume-test-name
```

But the creation of PVC not completed because I think it searches for a PV with this name and it doesn't exist. I commented the line with `volumeName` and this time PVC is created with PV with a random name. I am not completely sure but if `volumeName` config is set, in the code we first should create a PV with [`claimRef`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#persistentvolumespec-v1-core) to the PVC and when PVC is created, it will be bound it.

**Comment by consideRatio** [comment]
@bitnik ah I suspected that may be the case, thank you for clarifying this, then we should perhaps simply close this issue?

---

Regarding the z2jh project using kubespawner, if we would ignore backward compatebility, a StatefulSet should had been utilized instead of a deployment, see the discussion in this thread for more details about this: https://github.com/helm/charts/issues/1863#issuecomment-421327028

I've considered proposing to make the z2jh's hub Deployment a StatefulSet, but remaining backward compatible while doing it felt like too much trouble. Closing this issue for now as the `volumeName` solution won't do as clarified by @bitnik's feedback!

**Comment by hans-permana** [comment]
while this issue can be closed, my problem is still unsolved even when StatefulSet is used instead of Deployment. When a cluster is re-created, all the PVs and PVCs are gone. The solution to this is I think to create a PV with a specified `pdName` if not yet exists. This has to be done before the creation of the PVC. What do you think @consideRatio ?

**Comment by consideRatio** [comment]
Hmmm, I'm not confident about what happens if you create the full cluster itself, but if you delete the namespace and the helm release etc, then I know the PVs created by the PVCs can be undeleted and things will can be fine. For that, see:
1. https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/CHANGELOG.md#1-optional-ensure-the-hubs-and-users-data-isnt-lost
2. https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/CHANGELOG.md#troubleshooting---make-released-pvs-available-for-reuse

I'm curious to learn more about how a PV is coupled to a cluster, and if you could transfer it to another cluster, etc. I've only been using GKE, and never done such thing.
